### PR TITLE
AVRO-1818 - Avoid buffer copy in DeflateCodec.compress and decompress

### PR DESCRIPTION
**Pull request by nandorKollar**

### CODE REVIEW

File: lang/java/avro/src/main/java/org/apache/avro/file/DeflateCodec.java [hunk 1]
```diff
@@ -51,6 +53,18 @@ protected Codec createInstance() {
     }
   }
 
+  private static class Buffer extends ByteArrayOutputStream {
+    public synchronized byte[] getData() { return buf; }
```
Comment (gszadovszky) [inline lang/java/avro/src/main/java/org/apache/avro/file/DeflateCodec.java]: synchronized is useless here: returning a reference is atomic.

File: lang/java/trevni/core/src/main/java/org/apache/trevni/DeflateCodec.java [hunk 2]
```diff
@@ -32,24 +34,41 @@
   private Deflater deflater;
   private Inflater inflater;
 
+  private static class Buffer extends ByteArrayOutputStream {
+    public synchronized byte[] getData() { return buf; }
```
Comment (gszadovszky) [inline lang/java/trevni/core/src/main/java/org/apache/trevni/DeflateCodec.java]: Same as above.

File: lang/java/avro/src/main/java/org/apache/avro/file/DeflateCodec.java [hunk 3]
```diff
@@ -81,16 +95,15 @@ public ByteBuffer decompress(ByteBuffer data) throws IOException {
...
 
   private void writeAndClose(ByteBuffer data, OutputStream to) throws IOException {
-    byte[] input = data.array();
-    int offset = data.arrayOffset() + data.position();
-    int length = data.remaining();
     try {
-      to.write(input, offset, length);
+      WritableByteChannel channel = Channels.newChannel(to);
+
+      channel.write(data);
```
Comment (tomwhite) [inline lang/java/avro/src/main/java/org/apache/avro/file/DeflateCodec.java]: How is this change related to the bugfix?

Comment (tomwhite) [inline lang/java/avro/src/main/java/org/apache/avro/file/DeflateCodec.java]: @nandorKollar would it be possible to not include this change to address this issue?
